File: Sources/PowerpointViewer/SymbolFont.swift
import Foundation

/// Maps symbol-font bullet characters (Wingdings, Symbol) to Unicode.
///
/// PowerPoint stores list bullets as a raw character plus a symbol font, e.g.
/// Wingdings `v`. Rendering that character in the symbol font on macOS produces
/// glyphs that differ from what PowerPoint/Keynote show (Wingdings `v` becomes a
/// cluster of diamonds rather than the expected ❖). We translate the common
/// bullet characters to their Unicode equivalents and render them in the normal
/// text font, so bullets match the source presentation.
enum SymbolFont {
    /// Wingdings character → Unicode bullet glyph (common bullet subset).
    private static let wingdings: [Character: String] = [
        "l": "\u{25CF}",   // ● black circle
        "n": "\u{25A0}",   // ■ black square
        "o": "\u{25A1}",   // □ white square
        "p": "\u{2751}",   // ❑ lower-right shadowed white square
        "q": "\u{2752}",   // ❒ upper-right shadowed white square
        "u": "\u{2666}",   // ◆ black diamond
        "v": "\u{2756}",   // ❖ black diamond minus white X
        "w": "\u{2726}",   // ✦ black four-pointed star
        "\u{00A7}": "\u{25AA}", // ▪ small black square
        "\u{00A8}": "\u{25AB}", // ▫ small white square
        "\u{00D8}": "\u{27A2}", // ➢ three-D top-lit arrowhead
        "\u{00FC}": "\u{2714}", // ✔ check mark
        "\u{00FE}": "\u{2717}"  // ✗ ballot X
    ]

    /// Symbol-font (math) character → Unicode.
    private static let symbol: [Character: String] = [
        "\u{00B7}": "\u{2022}", // · → • bullet
        "\u{00A7}": "\u{2666}"  // diamond
    ]

    /// Returns a Unicode bullet for a symbol-font character, or nil to keep the
    /// raw character in its original font.
    static func unicodeBullet(char: String, font: String?) -> String? {
        guard let first = char.first else { return nil }
        switch font?.lowercased() {
        case .some(let f) where f.hasPrefix("wingdings"):
            return wingdings[first]
        case "symbol":
            return symbol[first]
        default:
            return nil
        }
    }
}

File: Windows/SlideViewer/Parsing/SymbolFont.cs
namespace SlideViewer.Parsing;

/// <summary>
/// Maps symbol-font bullet characters (Wingdings, Symbol) to Unicode.
///
/// PowerPoint stores bullets as a raw character plus a symbol font, e.g.
/// Wingdings 'v'. Rendering that character in the symbol font produces glyphs
/// that differ from what PowerPoint shows (Wingdings 'v' becomes a cluster of
/// diamonds instead of the expected ❖). Translating to Unicode and rendering in
/// the normal text font matches the source presentation on any platform.
/// </summary>
public static class SymbolFont
{
    private static readonly Dictionary<char, string> Wingdings = new()
    {
        ['l'] = "●",   // ● black circle
        ['n'] = "■",   // ■ black square
        ['o'] = "□",   // □ white square
        ['p'] = "❑",   // ❑ shadowed square
        ['q'] = "❒",   // ❒ shadowed square
        ['u'] = "♦",   // ◆ black diamond
        ['v'] = "❖",   // ❖ black diamond minus white X
        ['w'] = "✦",   // ✦ black four-pointed star
        ['§'] = "▪",  // ▪ small black square
        ['¨'] = "▫",  // ▫ small white square
        ['Ø'] = "➢",  // ➢ arrowhead
        ['ü'] = "✔",  // ✔ check mark
        ['þ'] = "✗",  // ✗ ballot X
    };

    private static readonly Dictionary<char, string> Symbol = new()
    {
        ['·'] = "•",  // • bullet
        ['§'] = "♦",  // ◆ diamond
    };

    /// <summary>Unicode bullet for a symbol-font character, or null to keep the
    /// raw character in its original font.</summary>
    public static string? UnicodeBullet(string ch, string? font)
    {
        if (string.IsNullOrEmpty(ch) || font == null) return null;
        var f = font.ToLowerInvariant();
        var c = ch[0];
        if (f.StartsWith("wingdings")) return Wingdings.TryGetValue(c, out var w) ? w : null;
        if (f == "symbol") return Symbol.TryGetValue(c, out var s) ? s : null;
        return null;
    }
}
